lib/readline.js

```diff
@@ -454,16 +454,27 @@ Interface.prototype._normalWrite = function(b) {
   }
 };
 
+// Count the Unicode code points in a string.
+function codePointLen(str) {
+  let count = 0;
+  // eslint-disable-next-line no-unused-vars
+  for (const char of str) {
+    count++;
+  }
+  return count;
+}
```

Comment:
I just had another look at this and I found the reason why I originally thought that `_insertString()` would break:

This function returns the number of characters and not the length. So I thought the cursor would from now on only increase by the number of characters and that would break lots of things. However, this function will mostly receive strings containing a single `UTF-16` code point. Therefore the insertion still works as expected and no corruption occurs. Only for tab completion a longer string would be inserted and the REPLs completer does not seem to handle completions for such code points.

To conclude: this function is not needed and the code in `_insertString()` should be kept as it was.

Reply:
@BridgeAR What about pasting? Doesn't `_insertString()` handle pastes, which may include multiple possibly astral characters.

Reply:
Yes and no. All data (also pastes) is streamed and then a for loop iterates over each UTF-16 character (https://github.com/nodejs/node/blob/200560cc0decbec35d9188ac9e155ee5bb7381e0/lib/readline.js#L1105). That data is then handled in an emitKeys and causes keypress events to be emitted (https://github.com/nodejs/node/blob/200560cc0decbec35d9188ac9e155ee5bb7381e0/lib/internal/readline.js#L427). Each event triggers the onkeypress function which triggers the `_ttywrite` (https://github.com/nodejs/node/blob/200560cc0decbec35d9188ac9e155ee5bb7381e0/lib/readline.js#L184) with that single character.

So astral characters will be broken into two characters and handled properly.

However, if that would not be the case, the `_insertString()` function would cause issues (the cursor would be set to a lower value than the string length if it contains astral characters). That is the reason why I originally asked for further tests and thought things might break.